rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,295 @@
+    /// option on ReadOptions)
+    ///
+    /// default: uses the clock time configured in the DbOptions
+    pub timestamp_ms: Option<i64>,
+   
+    /// the time-to-live (ttl) for this insertion. If this insert overwrites an existing
+    /// database entry, the TTL for the most recent entry will be canonical.
+    ///
+    /// default: the TTL configured in DbOptions when opening a SlateDB session
+    pub ttl: Option<Duration>,
+}
+```
+
+`Db#get_with_options` can optionally specify: a grace period with which to continue searching the
+LSM tree for a more recent insertion. This allows you to specify an upper bound for what you expect
+out-of-order insertions to be.
+
+```rust
+pub struct ReadOptions {
+    // ...
+
+    /// if set to a positive number, SlateDB will not return the first matching row for
+    /// each key lookup and instead will continue to traverse the LSM tree until the next
+    /// matching key is older than the found key's timestamp + retrieval_grace_ms. Logically,
+    /// this is an "upper bound" on the magnitude of out-of-order insertions.
+    ///
+    /// Example: if there are three insertions [KeyA@100, KeyA@50, KeyA@150] that are each
+    /// in separate layers of the LSM tree, you will see the following behavior:
+    /// - if retrieval_grace_ms is <=0, KeyA@100 will be returned, and only the first SST
+    ///   will be scanned
+    /// - if retrieval_grace_ms is <=50, KeyA@100 will be returned after having scanned
+    ///   the SST that contains KeyA@50
+    /// - if retrieval_grace_ms is > 50 and <= 100, KeyA@150 will be returned after having
+    ///   scanned the SST containing KeyA@150
+    /// - otherwise, KeyA@150 will be returned after scanning all levels of the LSM tree
+    ///
+    /// default: -1 (the first matching key will be retrieved)
+    pub retrieval_grace_ms: i64,
+}
+```
+
+**Callout for TTL**: I have decided that for this RFC we will not introduce the ability to disable
+filtering out TTL-expired records on retrieval. The default implementation will read the metadata
+for every retrieved row and filter out records where `ts + tll < now`. We can, as a future
+optimization, enable lazy deserialization of the metadata so that this doesn't need to happen on
+each row retrieved. This can be part of the API design for compaction filters (
+see https://github.com/slatedb/slatedb/issues/225)
+
+**Alternative**: if this retrieval grace concept is too complicated, we can also consider specifying
+a "base timestamp", that will retrieve the first result >= the configured base timestamp, or the 
+most recent one if none are greater than or equal to the base timestamp. This is significantly less
+flexible, however, as many rows may not be updated in a long time (though most of those rows will be
+in the base layer of the LSM tree anyway, so it may not be so bad).
+
+### SlateDB::Get
+
+SlateDB currently returns a `Option<Bytes>` on a successful read operation. Now that additional
```

Comment:
What are some use cases for wanting to see the metadata bytes?

Reply:
To get the timestamp associated with the record (that seems helpful?)